pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java

```diff
@@ -49,84 +50,80 @@
   boolean emptyResult = false;
   private BrokerRequest brokerRequest;
 
   public StarTreeIndexOperator(IndexSegment segment, BrokerRequest brokerRequest) {
     this.segment = segment;
     this.brokerRequest = brokerRequest;
-    eligibleEqualityPredicatesMap = new HashMap<>();
-    inEligiblePredicatesMap = new HashMap<>();
-    nonEqualityPredicatesMap = new HashMap<>();
+    equalityPredicateColumns = new HashSet<>();
+    groupByColumns = new HashSet<>();
+    predicatesMap = new HashMap<>();
     initPredicatesToEvaluate();
   }
 
   private void initPredicatesToEvaluate() {
     FilterQueryTree filterTree = RequestUtils.generateFilterQueryTree(brokerRequest);
-    //find all filter columns
+    // Find all filter columns
     if (filterTree != null) {
       if (filterTree.getChildren() != null && !filterTree.getChildren().isEmpty()) {
         for (FilterQueryTree childFilter : filterTree.getChildren()) {
-          //nested filters are not supported
+          // Nested filters are not supported
           assert childFilter.getChildren() == null || childFilter.getChildren().isEmpty();
           processFilterTree(childFilter);
         }
       } else {
         processFilterTree(filterTree);
       }
     }
-    //group by columns, we cannot lose group by columns during traversal
+    // Group by columns, we cannot lose group by columns during traversal
     GroupBy groupBy = brokerRequest.getGroupBy();
     if (groupBy != null) {
-      for (String groupByCol : groupBy.getColumns()) {
-        //remove if there was a filter predicate set on this column in the query
-        if (eligibleEqualityPredicatesMap.containsKey(groupByCol)) {
-          inEligiblePredicatesMap.put(groupByCol, eligibleEqualityPredicatesMap.get(groupByCol));
-          eligibleEqualityPredicatesMap.remove(groupByCol);
-        } else {
-          //there is no predicate but we cannot lose this dimension while traversing
-          inEligiblePredicatesMap.put(groupByCol, null);
-        }
-      }
+      groupByColumns.addAll(groupBy.getColumns());
     }
   }
 
   private void processFilterTree(FilterQueryTree childFilter) {
     String column = childFilter.getColumn();
-    //only equality predicates are supported
+    // Only equality predicates are supported
     Predicate predicate = Predicate.newPredicate(childFilter);
     Dictionary dictionary = segment.getDataSource(column).getDictionary();
+    PredicateEntry predicateEntry = null;
     if (childFilter.getOperator() == FilterOperator.EQUALITY) {
       EqPredicate eqPredicate = (EqPredicate) predicate;
-      //computing dictionaryId allows us early termination and avoids multiple looks up during tree traversal
+      // Computing dictionaryId allows us early termination and avoids multiple looks up during tree
+      // traversal
       int dictId = dictionary.indexOf(eqPredicate.getEqualsValue());
       if (dictId < 0) {
-        //empty result
+        // Empty result
         emptyResult = true;
```

Comment:
When we set emptyResult to true, should we just return because all other operations are redundant?